domain/application/state/exposed.go

```diff
@@ -5,38 +5,481 @@ package state
+		if endpoint.Name.Valid {
+			endpointName = endpoint.Name.String
+		}
+		exposed[endpointName] = application.ExposedEndpoint{
+			ExposeToCIDRs:    set.NewStrings(),
+			ExposeToSpaceIDs: set.NewStrings(),
+		}
+	}
+
+	for _, endpointCIDR := range cidrs {
+		endpointName := wildcardEndpointName
+		if endpointCIDR.EndpointName.Valid {
+			endpointName = endpointCIDR.EndpointName.String
+		}
+		exposed[endpointName].ExposeToCIDRs.Add(endpointCIDR.CIDR)
+	}
+	for _, endpointSpace := range spaces {
+		endpointName := wildcardEndpointName
+		if endpointSpace.EndpointName.Valid {
+			endpointName = endpointSpace.EndpointName.String
+		}
+		exposed[endpointName].ExposeToSpaceIDs.Add(endpointSpace.SpaceUUID)
+	}
+	return exposed
 }
 
 // UnsetExposeSettings removes the expose settings for the provided list of
 // endpoint names. If the resulting exposed endpoints map for the application
 // becomes empty after the settings are removed, the application will be
 // automatically unexposed.
-func (s *State) UnsetExposeSettings(ctx context.Context, appID coreapplication.ID, exposedEndpoints set.Strings) error {
-	return nil
+func (st *State) UnsetExposeSettings(ctx context.Context, appID coreapplication.ID, exposedEndpoints set.Strings) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		for _, endpoint := range exposedEndpoints.Values() {
+			if err := st.unsetExposedEndpoint(ctx, tx, appID, endpoint); err != nil {
+				return errors.Capture(err)
+			}
+		}
+		return nil
+	})
+
+	return errors.Capture(err)
 }
 
 // MergeExposeSettings marks the application as exposed and merges the provided
 // ExposedEndpoint details into the current set of expose settings. The merge
 // operation will overwrite expose settings for each existing endpoint name.
-func (s *State) MergeExposeSettings(ctx context.Context, appID coreapplication.ID, exposedEndpoints map[string]application.ExposedEndpoint) error {
+func (st *State) MergeExposeSettings(ctx context.Context, appID coreapplication.ID, exposedEndpoints map[string]application.ExposedEndpoint) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		for endpoint, exposedEndpoint := range exposedEndpoints {
+			if err := st.unsetExposedEndpoint(ctx, tx, appID, endpoint); err != nil {
+				return errors.Capture(err)
+			}
+			if err := st.upsertExposedCIDRs(ctx, tx, appID, endpoint, exposedEndpoint.ExposeToCIDRs); err != nil {
+				return errors.Capture(err)
+			}
+			if err := st.upsertExposedSpaces(ctx, tx, appID, endpoint, exposedEndpoint.ExposeToSpaceIDs); err != nil {
+				return errors.Capture(err)
+			}
+		}
+		return nil
+	})
+
+	return errors.Capture(err)
+}
+
+func (st *State) unsetExposedEndpoint(ctx context.Context, tx *sqlair.TX, appID coreapplication.ID, endpoint string) error {
```

Comment:
Break this up, so it's the following:

```suggestion
func (st *State) unsetExposedEndpoint(ctx context.Context, tx *sqlair.TX, appID coreapplication.ID, endpoint string) error {
     ... st.unsetExposedEndpointCIDR(...)
     ... st.unsetExposedEndpoingSpace(...)
```

Reply:
Good call, done.